[
    {
        "content": "<p>From: Lukas Bulwahn &lt;<a href=\"mailto:lukas.bulwahn@gmail.com\">lukas.bulwahn@gmail.com</a>&gt;<br>\nHi all,</p>\n<p>as information for further analysis, I can confirm that this issue<br>\nalso occurs on the recent development version of Isabelle with hg id:<br>\n9c31678d2139.</p>\n<p>Lukas</p>",
        "id": 294730312,
        "sender_full_name": "Email Gateway",
        "timestamp": 1661187775
    },
    {
        "content": "<p>From: Makarius &lt;<a href=\"mailto:makarius@sketis.net\">makarius@sketis.net</a>&gt;<br>\nOn 06/04/18 10:00, Peter Zeller wrote:</p>\n<blockquote>\n<p>[I hope this is the right place for bug reports -- I did not find any<br>\npublic issue tracker.]</p>\n</blockquote>\n<p>Yes, definitely.</p>\n<p>The isabelle-users mailing list is for all kinds of observations and<br>\ndiscussions of official Isabelle releases. Likewise the isabelle-dev<br>\nmailing list is for arbitrary repository snapshots (with indication of<br>\nthe changeset id) and in particular the administrative process around it.</p>\n<p>Thus as a user of official Isabelle2017, anything posted on<br>\nisabelle-users is the correct place.</p>\n<blockquote>\n<p>When trying to use the smt method I am getting the error: exception<br>\nOption raised (line 82 of \"General/basics.ML\")</p>\n<p>Here is a minimized example where this happens:</p>\n<p>lemma<br>\n      assumes \"⋀a . ∃!x. f x = [a]\"<br>\n      shows \"∃!x. ∃!y. f x = [a] ∧ f y = [b]\"<br>\n   proof  -<br>\n      have ?thesis<br>\n        using assms by smt (* exception Option raised (line 82 of<br>\n   \"General/basics.ML\") *)<br>\n      show ?thesis<br>\n        using assms by metis (* works *)<br>\n   qed</p>\n<p>I am using Isabelle 2017 on Linux.</p>\n</blockquote>\n<p>Thanks for this clear example. It looks like a problem in the<br>\npost-processing of Z3 proofs by the \"smt\" proof method, but could be<br>\nalso in Z3 itself. Included is an exception trace of the situation. The<br>\nexception is raised here<br>\n<a href=\"http://isabelle.in.tum.de/repos/isabelle/file/Isabelle2017/src/HOL/Tools/SMT/z3_proof.ML#l221\">http://isabelle.in.tum.de/repos/isabelle/file/Isabelle2017/src/HOL/Tools/SMT/z3_proof.ML#l221</a></p>\n<p>val objTs = map (the o Symtab.lookup env) bounds</p>\n<p>with the following run-time values:</p>\n<p>bounds = [\"?1\", \"?0\"]<br>\n  env = {(\"?0\", \"'b\")}</p>\n<p>The module Z3_Proof is by Sascha Böhme -- I hope he will have an idea<br>\nwhat is wrong.</p>\n<blockquote>\n<p>The error message is not very helpful, since the \"the\" function is<br>\ncalled in many places. If I want to understand this bug, what would be<br>\nthe best way to debug it and find the problematic call to \"the\"?</p>\n</blockquote>\n<p>By default, Poly/ML provides hardly any exception trace information --<br>\nfor performance reasons. In order to figure out the above, I did the<br>\nfollowing:</p>\n<p>* Remove the existing HOL heap image (presumably in<br>\n$ISABELLE_HOME/heaps) and rebuild it with debugging enabled globally<br>\n(this can be slow):</p>\n<p>isabelle build -o ML_debugger -b HOL -s</p>\n<p>The option -s merely accommodates the typical situation for end-users<br>\nthat ML heaps are built automatically inside the application directory<br>\n$ISABELLE_HOME, not the configuration directory $ISABELLE_HOME_USER.</p>\n<p>* Start Isabelle/jEdit with session \"HOL\" and run the above example in<br>\na context with declare [[ML_exception_debugger]] -- it produces a full<br>\nexception trace for all modules that have been compiled with ML_debugger<br>\nenabled.</p>\n<p>* Optional: restarted Isabelle/jEdit with session \"Pure\" and<br>\nexperiment directly with the relevant sources, limiting the debugging<br>\ninformation to the relevant module. I also tried the Isabelle/ML<br>\nsource-level debugger IDE, but it turned out too awkward at this point.<br>\nUltimately, I merely put in some print statements in the proper place:<br>\nsee also the include changeset \"ch\" and Scratch.thy for the record.</p>\n<p>Such tinkering with Isabelle/ML sources and the Poly/ML runtime<br>\nenvironment can be quite tedious. The above is merely for information,<br>\nfor anybody who is interested in that.</p>\n<p>When seeing a problem, there is no moral obligation to figure out all<br>\nits details. It is sufficient to present a clear example as you have<br>\ndone above.</p>\n<p>Makarius<br>\n<a href=\"/user_uploads/14278/ZI2a6X-xEm7xdrsEnENllYun/exception_trace\">exception_trace</a><br>\n<a href=\"/user_uploads/14278/3Kt9Tjo9FTp6wuV5rUX4EekG/ch\">ch</a><br>\n<a href=\"/user_uploads/14278/WB43-bHT13QJmWicRGkSo-Lh/Scratch.thy\">Scratch.thy</a></p>",
        "id": 294731050,
        "sender_full_name": "Email Gateway",
        "timestamp": 1661188067
    },
    {
        "content": "<p>From: Sascha Böhme &lt;<a href=\"mailto:boehmes@in.tum.de\">boehmes@in.tum.de</a>&gt;<br>\nHello,</p>\n<p>The exception is caused by a problem in the proof trace produced by Z3. I’ve slightly improved the error reporting issued by the smt method. This will be contained in next Isabelle release.</p>\n<p>I’m sorry to say that the smt method might fail to prove such goals. Prefer metis for tough problems that involve quantifiers.</p>\n<p>Cheers,<br>\nSascha</p>\n<p>Von: Makarius<br>\nGesendet: Mittwoch, 25. April 2018 14:36<br>\nAn: Peter Zeller; <a href=\"mailto:cl-isabelle-users@lists.cam.ac.uk\">cl-isabelle-users@lists.cam.ac.uk</a><br>\nCc: Sascha Boehme<br>\nBetreff: Re: [isabelle] Bug report: Exception Option raised when using smtmethod</p>\n<p>On 06/04/18 10:00, Peter Zeller wrote:</p>\n<blockquote>\n<p>[I hope this is the right place for bug reports -- I did not find any<br>\npublic issue tracker.]</p>\n</blockquote>\n<p>Yes, definitely.</p>\n<p>The isabelle-users mailing list is for all kinds of observations and<br>\ndiscussions of official Isabelle releases. Likewise the isabelle-dev<br>\nmailing list is for arbitrary repository snapshots (with indication of<br>\nthe changeset id) and in particular the administrative process around it.</p>\n<p>Thus as a user of official Isabelle2017, anything posted on<br>\nisabelle-users is the correct place.</p>\n<blockquote>\n<p>When trying to use the smt method I am getting the error: exception<br>\nOption raised (line 82 of \"General/basics.ML\")</p>\n<p>Here is a minimized example where this happens:</p>\n<p>lemma<br>\n      assumes \"⋀a . ∃!x. f x = [a]\"<br>\n      shows \"∃!x. ∃!y. f x = [a] ∧ f y = [b]\"<br>\n   proof  -<br>\n      have ?thesis<br>\n        using assms by smt (* exception Option raised (line 82 of<br>\n   \"General/basics.ML\") *)<br>\n      show ?thesis<br>\n        using assms by metis (* works *)<br>\n   qed</p>\n<p>I am using Isabelle 2017 on Linux.</p>\n</blockquote>\n<p>Thanks for this clear example. It looks like a problem in the<br>\npost-processing of Z3 proofs by the \"smt\" proof method, but could be<br>\nalso in Z3 itself. Included is an exception trace of the situation. The<br>\nexception is raised here<br>\n<a href=\"http://isabelle.in.tum.de/repos/isabelle/file/Isabelle2017/src/HOL/Tools/SMT/z3_proof.ML#l221\">http://isabelle.in.tum.de/repos/isabelle/file/Isabelle2017/src/HOL/Tools/SMT/z3_proof.ML#l221</a></p>\n<p>val objTs = map (the o Symtab.lookup env) bounds</p>\n<p>with the following run-time values:</p>\n<p>bounds = [\"?1\", \"?0\"]<br>\n  env = {(\"?0\", \"'b\")}</p>\n<p>The module Z3_Proof is by Sascha Böhme -- I hope he will have an idea<br>\nwhat is wrong.</p>\n<blockquote>\n<p>The error message is not very helpful, since the \"the\" function is<br>\ncalled in many places. If I want to understand this bug, what would be<br>\nthe best way to debug it and find the problematic call to \"the\"?</p>\n</blockquote>\n<p>By default, Poly/ML provides hardly any exception trace information --<br>\nfor performance reasons. In order to figure out the above, I did the<br>\nfollowing:</p>\n<p>* Remove the existing HOL heap image (presumably in<br>\n$ISABELLE_HOME/heaps) and rebuild it with debugging enabled globally<br>\n(this can be slow):</p>\n<p>isabelle build -o ML_debugger -b HOL -s</p>\n<p>The option -s merely accommodates the typical situation for end-users<br>\nthat ML heaps are built automatically inside the application directory<br>\n$ISABELLE_HOME, not the configuration directory $ISABELLE_HOME_USER.</p>\n<p>* Start Isabelle/jEdit with session \"HOL\" and run the above example in<br>\na context with declare [[ML_exception_debugger]] -- it produces a full<br>\nexception trace for all modules that have been compiled with ML_debugger<br>\nenabled.</p>\n<p>* Optional: restarted Isabelle/jEdit with session \"Pure\" and<br>\nexperiment directly with the relevant sources, limiting the debugging<br>\ninformation to the relevant module. I also tried the Isabelle/ML<br>\nsource-level debugger IDE, but it turned out too awkward at this point.<br>\nUltimately, I merely put in some print statements in the proper place:<br>\nsee also the include changeset \"ch\" and Scratch.thy for the record.</p>\n<p>Such tinkering with Isabelle/ML sources and the Poly/ML runtime<br>\nenvironment can be quite tedious. The above is merely for information,<br>\nfor anybody who is interested in that.</p>\n<p>When seeing a problem, there is no moral obligation to figure out all<br>\nits details. It is sufficient to present a clear example as you have<br>\ndone above.</p>\n<p>Makarius</p>",
        "id": 294731527,
        "sender_full_name": "Email Gateway",
        "timestamp": 1661188229
    },
    {
        "content": "<p>From: Peter Zeller &lt;<a href=\"mailto:p_zeller@cs.uni-kl.de\">p_zeller@cs.uni-kl.de</a>&gt;<br>\nHello,</p>\n<p>[I hope this is the right place for bug reports -- I did not find any <br>\npublic issue tracker.]</p>\n<p>When trying to use the smt method I am getting the error: exception <br>\nOption raised (line 82 of \"General/basics.ML\")</p>\n<p>Here is a minimized example where this happens:</p>\n<p>lemma<br>\n       assumes \"⋀a . ∃!x. f x = [a]\"<br>\n       shows \"∃!x. ∃!y. f x = [a] ∧ f y = [b]\"<br>\n    proof  -<br>\n       have ?thesis<br>\n         using assms by smt (* exception Option raised (line 82 of<br>\n    \"General/basics.ML\") *)<br>\n       show ?thesis<br>\n         using assms by metis (* works *)<br>\n    qed</p>\n<p>I am using Isabelle 2017 on Linux.</p>\n<p>The error message is not very helpful, since the \"the\" function is <br>\ncalled in many places. If I want to understand this bug, what would be <br>\nthe best way to debug it and find the problematic call to \"the\"?</p>\n<p>Regards,<br>\nPeter<br>\n<a href=\"/user_uploads/14278/LGtDy0D3OOmuRqkkjFj98BkU/Scratch.thy\">Scratch.thy</a></p>",
        "id": 294731692,
        "sender_full_name": "Email Gateway",
        "timestamp": 1661188299
    }
]